// auth0-deploy-cli/lib/tools/auth0/handlers/guardianFactors.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.schema = void 0;
const default_1 = __importDefault(require("./default"));
const constants_1 = __importDefault(require("../../constants"));
exports.schema = {
    type: 'array',
    items: {
        type: 'object',
        properties: {
            name: { type: 'string', enum: constants_1.default.GUARDIAN_FACTORS },
        },
        required: ['name'],
    },
};
class GuardianFactorsHandler extends default_1.default {
    constructor(options) {
        super(Object.assign(Object.assign({}, options), { type: 'guardianFactors', id: 'name' }));
    }
    getType() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.existing)
                return this.existing;
            this.existing = yield this.client.guardian.getFactors();
            return this.existing;
        });
    }
    processChanges(assets) {
        return __awaiter(this, void 0, void 0, function* () {
            // No API to delete or create guardianFactors, we can only update.
            const { guardianFactors } = assets;
            // Do nothing if not set
            if (!guardianFactors || !guardianFactors.length)
                return;
            // Process each factor
            yield Promise.all(guardianFactors.map((factor) => __awaiter(this, void 0, void 0, function* () {
                const data = Object.assign({}, factor);
                const params = { name: factor.name };
                delete data.name;
                yield this.client.guardian.updateFactor(params, data);
                this.didUpdate(params);
                this.updated += 1;
            })));
        });
    }
}
exports.default = GuardianFactorsHandler;
