// auth0-deploy-cli/lib/tools/auth0/handlers/guardianPhoneFactorMessageTypes.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.schema = void 0;
const default_1 = __importDefault(require("./default"));
const constants_1 = __importDefault(require("../../constants"));
exports.schema = {
    type: 'object',
    properties: {
        message_types: {
            type: 'array',
            items: {
                type: 'string',
                enum: constants_1.default.GUARDIAN_PHONE_MESSAGE_TYPES,
            },
        },
    },
    additionalProperties: false,
};
const isFeatureUnavailableError = (err) => {
    if (err.statusCode === 404) {
        // Older Management API version where the endpoint is not available.
        return true;
    }
    if (err.statusCode === 403 &&
        err.originalError &&
        err.originalError.response &&
        err.originalError.response.body &&
        err.originalError.response.body.errorCode === 'voice_mfa_not_allowed') {
        // Recent Management API version, but with feature explicitly disabled.
        return true;
    }
    return false;
};
class GuardianPhoneMessageTypesHandler extends default_1.default {
    constructor(options) {
        super(Object.assign(Object.assign({}, options), { type: 'guardianPhoneFactorMessageTypes' }));
    }
    getType() {
        return __awaiter(this, void 0, void 0, function* () {
            // in case client version does not support the operation
            if (!this.client.guardian ||
                typeof this.client.guardian.getPhoneFactorMessageTypes !== 'function') {
                return {};
            }
            if (this.existing)
                return this.existing;
            try {
                this.existing = yield this.client.guardian.getPhoneFactorMessageTypes();
            }
            catch (e) {
                if (isFeatureUnavailableError(e)) {
                    // Gracefully skip processing this configuration value.
                    return {};
                }
                throw e;
            }
            return this.existing;
        });
    }
    processChanges(assets) {
        return __awaiter(this, void 0, void 0, function* () {
            // No API to delete or create guardianPhoneFactorMessageTypes, we can only update.
            const { guardianPhoneFactorMessageTypes } = assets;
            // Do nothing if not set
            if (!guardianPhoneFactorMessageTypes || !guardianPhoneFactorMessageTypes.message_types)
                return;
            const params = {};
            const data = guardianPhoneFactorMessageTypes;
            yield this.client.guardian.updatePhoneFactorMessageTypes(params, data);
            this.updated += 1;
            this.didUpdate(guardianPhoneFactorMessageTypes);
        });
    }
}
exports.default = GuardianPhoneMessageTypesHandler;
