// auth0-deploy-cli/lib/tools/auth0/handlers/guardianFactorTemplates.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.schema = void 0;
const default_1 = __importDefault(require("./default"));
const constants_1 = __importDefault(require("../../constants"));
exports.schema = {
    type: 'array',
    items: {
        type: 'object',
        properties: {
            name: { type: 'string', enum: constants_1.default.GUARDIAN_FACTOR_TEMPLATES },
        },
        required: ['name'],
    },
};
class GuardianFactorTemplatesHandler extends default_1.default {
    constructor(options) {
        super(Object.assign(Object.assign({}, options), { type: 'guardianFactorTemplates', id: 'name' }));
    }
    getType() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.existing)
                return this.existing;
            const data = yield Promise.all(constants_1.default.GUARDIAN_FACTOR_TEMPLATES.map((name) => __awaiter(this, void 0, void 0, function* () {
                const templates = yield this.client.guardian.getFactorTemplates({ name });
                return Object.assign({ name }, templates);
            })));
            // Filter out empty, should have more then 1 keys (name)
            return data.filter((d) => Object.keys(d).length > 1);
        });
    }
    processChanges(assets) {
        return __awaiter(this, void 0, void 0, function* () {
            // No API to delete or create guardianFactorTemplates, we can only update.
            const { guardianFactorTemplates } = assets;
            // Do nothing if not set
            if (!guardianFactorTemplates || !guardianFactorTemplates.length)
                return;
            // Process each factor templates
            yield Promise.all(guardianFactorTemplates.map((fatorTemplates) => __awaiter(this, void 0, void 0, function* () {
                const data = Object.assign({}, fatorTemplates);
                const params = { name: fatorTemplates.name };
                delete data.name;
                yield this.client.guardian.updateFactorTemplates(params, data);
                this.didUpdate(params);
                this.updated += 1;
            })));
        });
    }
}
exports.default = GuardianFactorTemplatesHandler;
