// auth0-deploy-cli/lib/tools/auth0/handlers/guardianPolicies.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.schema = void 0;
const default_1 = __importDefault(require("./default"));
const constants_1 = __importDefault(require("../../constants"));
exports.schema = {
    type: 'object',
    properties: {
        policies: {
            type: 'array',
            items: {
                type: 'string',
                enum: constants_1.default.GUARDIAN_POLICIES,
            },
        },
    },
    additionalProperties: false,
};
class GuardianPoliciesHandler extends default_1.default {
    constructor(options) {
        super(Object.assign(Object.assign({}, options), { type: 'guardianPolicies' }));
    }
    //TODO: standardize empty object literal with more intentional empty indicator
    getType() {
        return __awaiter(this, void 0, void 0, function* () {
            // in case client version does not support the operation
            if (!this.client.guardian || typeof this.client.guardian.getPolicies !== 'function') {
                return {};
            }
            if (this.existing)
                return this.existing;
            const policies = yield this.client.guardian.getPolicies();
            this.existing = { policies };
            return this.existing;
        });
    }
    processChanges(assets) {
        return __awaiter(this, void 0, void 0, function* () {
            // No API to delete or create guardianPolicies, we can only update.
            const { guardianPolicies } = assets;
            // Do nothing if not set
            if (!guardianPolicies || !guardianPolicies.policies)
                return;
            const params = {};
            const data = guardianPolicies.policies;
            yield this.client.guardian.updatePolicies(params, data);
            this.updated += 1;
            this.didUpdate(guardianPolicies);
        });
    }
}
exports.default = GuardianPoliciesHandler;
