// auth0-deploy-cli/lib/tools/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Auth0 = exports.wrapArrayReplaceMarkersInQuotes = exports.loadFileAndReplaceKeywords = exports.keywordReplace = exports.deploy = exports.constants = void 0;
const constants_1 = __importDefault(require("./constants"));
exports.constants = constants_1.default;
const deploy_1 = __importDefault(require("./deploy"));
exports.deploy = deploy_1.default;
const auth0_1 = __importDefault(require("./auth0"));
exports.Auth0 = auth0_1.default;
const utils_1 = require("./utils");
Object.defineProperty(exports, "keywordReplace", { enumerable: true, get: function () { return utils_1.keywordReplace; } });
Object.defineProperty(exports, "loadFileAndReplaceKeywords", { enumerable: true, get: function () { return utils_1.loadFileAndReplaceKeywords; } });
Object.defineProperty(exports, "wrapArrayReplaceMarkersInQuotes", { enumerable: true, get: function () { return utils_1.wrapArrayReplaceMarkersInQuotes; } });
exports.default = {
    constants: constants_1.default,
    deploy: deploy_1.default,
    keywordReplace: utils_1.keywordReplace,
    loadFileAndReplaceKeywords: utils_1.loadFileAndReplaceKeywords,
    wrapArrayReplaceMarkersInQuotes: utils_1.wrapArrayReplaceMarkersInQuotes,
    Auth0: auth0_1.default,
};
